domain/annotation/state/state.go

```diff
@@ -0,0 +1,274 @@
+			return fmt.Errorf("unable to find UUID for ID: %q %w", id.Name, errors.NotFound)
+		}
+
+		uuid, ok := result["uuid"].(string)
+		if !ok {
+			return fmt.Errorf("unable to find UUID for ID: %q %w", id.Name, errors.NotFound)
+		}
+		// Querying for annotations
+		return tx.Query(ctx, getAnnotationsStmt, sqlair.M{
+			"uuid": uuid,
+		}).GetAll(&annotationsResults)
+	})
+
+	if err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// No errors, we return empty map if no annotation is found
+			return map[string]string{}, nil
+		}
+		return nil, errors.Annotatef(err, "loading annotations for ID: %q", id.Name)
+	}
+
+	annotations := make(map[string]string, len(annotationsResults))
+
+	for _, result := range annotationsResults {
+		annotations[result.Key] = result.Value
+	}
+
+	return annotations, errors.Trace(err)
+}
+
+// SetAnnotations associates key/value annotation pairs with a given ID.
+// If annotation already exists for the given ID, then it will be updated with
+// the given value. Setting a key's value to "" will remove the key from the annotations map
+// (functionally unsetting the key).
+func (st *State) SetAnnotations(ctx context.Context, id annotations.ID,
+	annotationsParam map[string]string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Separate the annotations that are to be setted vs removed
+	toUpsert := make(map[string]string)
+	var toRemove []string
+
+	for key, value := range annotationsParam {
+		if value == "" {
+			toRemove = append(toRemove, fmt.Sprintf("'%s'", key))
+		} else {
+			toUpsert[key] = value
+		}
+	}
+
+	if id.Kind == annotations.KindModel {
+		return st.setAnnotationsForModel(ctx, db, id, toUpsert, toRemove)
+	}
+	return st.setAnnotationsForID(ctx, db, id, toUpsert, toRemove)
+}
+
+// setAnnotationsForID associates key/value pairs with the given ID. This is separate from the
+// setAnnotationsForModel because for non-model ID Kinds we need to find the uuid of the id before
+// we add an annotation in the corresponding annotation table.
+func (st *State) setAnnotationsForID(ctx context.Context, db database.TxnRunner, id annotations.ID,
+	toUpsert map[string]string, toRemove []string) error {
+
+	// Prepare query for getting the UUID of id.
+	kindQuery, kindQueryParam, err := uuidQueryForID(id)
+	if err != nil {
+		return errors.Annotatef(err, "preparing uuid retrieval query for ID: %q", id.Name)
+	}
+	kindQueryStmt, err := sqlair.Prepare(kindQuery, sqlair.M{})
+	if err != nil {
+		return errors.Annotatef(err, "preparing uuid retrieval query for ID: %q", id.Name)
+	}
+
+	// Prepare query for inserting and deleting annotations for id
+	var setAnnotationsQuery string
+	var deleteAnnotationsQuery string
+
+	var setAnnotationsStmt *sqlair.Statement
+	var deleteAnnotationsStmt *sqlair.Statement
+
+	// Set query
+	setAnnotationsQuery = setAnnotationQueryForID(id)
+	// Delete query
+	deleteAnnotationsQuery = deleteAnnotationsQueryForID(id, strings.Join(toRemove, ", "))
```

Comment:
We have common functions for these sorts of things, see `internal/database/statement.go`. 

See:

 - SliceToPlaceholder
 - MakeBindArgs